In [1]:
import pandas as pd 
import json

In [2]:
### Reading the three jsons and loading then into a dataframe

# File paths
file_paths = {
    'brands': 'brands.json',
    'receipts': 'receipts.json',
    'users': 'users.json'
}

# Function to read and parse JSON lines file
def read_json_file(file_path):
    data = []
    with open(file_path, 'r') as file:
        for line in file:
            line = line.strip()
            if line:  
                try:
                    json_object = json.loads(line)  
                    data.append(json_object)  
                except json.JSONDecodeError as e:
                    print(f"Error decoding JSON on line: {line} - Error: {e}")
    return data

# Reading each file and converting to DataFrame
brands_data = read_json_file(file_paths['brands'])
receipts_data = read_json_file(file_paths['receipts'])
users_data = read_json_file(file_paths['users'])

# Converting lists of JSON objects to DataFrames
brands = pd.DataFrame(brands_data)
receipts  = pd.DataFrame(receipts_data)
users = pd.DataFrame(users_data)

In [3]:
receipts.head()

_id  bonusPointsEarned  \
0  {'$oid': '5ff1e1eb0a720f0523000575'}              500.0   
1  {'$oid': '5ff1e1bb0a720f052300056b'}              150.0   
2  {'$oid': '5ff1e1f10a720f052300057a'}                5.0   
3  {'$oid': '5ff1e1ee0a7214ada100056f'}                5.0   
4  {'$oid': '5ff1e1d20a7214ada1000561'}                5.0   

                             bonusPointsEarnedReason  \
0  Receipt number 2 completed, bonus point schedu...   
1  Receipt number 5 completed, bonus point schedu...   
2                         All-receipts receipt bonus   
3                         All-receipts receipt bonus   
4                         All-receipts receipt bonus   

                 createDate               dateScanned  \
0  {'$date': 1609687531000}  {'$date': 1609687531000}   
1  {'$date': 1609687483000}  {'$date': 1609687483000}   
2  {'$date': 1609687537000}  {'$date': 1609687537000}   
3  {'$date': 1609687534000}  {'$date': 1609687534000}   
4  {'$date': 1609687506000}  {'$date': 1609687506000}   

               finishedDate                modifyDate  \
0  {'$date': 1609687531000}  {'$date': 1609687536000}   
1  {'$date': 1609687483000}  {'$date': 1609687488000}   
2                       NaN  {'$date': 1609687542000}   
3  {'$date': 1609687534000}  {'$date': 1609687539000}   
4  {'$date': 1609687511000}  {'$date': 1609687511000}   

          pointsAwardedDate pointsEarned              purchaseDate  \
0  {'$date': 1609687531000}        500.0  {'$date': 1609632000000}   
1  {'$date': 1609687483000}        150.0  {'$date': 1609601083000}   
2                       NaN            5  {'$date': 1609632000000}   
3  {'$date': 1609687534000}          5.0  {'$date': 1609632000000}   
4  {'$date': 1609687506000}          5.0  {'$date': 1609601106000}   

   purchasedItemCount                             rewardsReceiptItemList  \
0                 5.0  [{'barcode': '4011', 'description': 'ITEM NOT ...   
1                 2.0  [{'barcode': '4011', 'description': 'ITEM NOT ...   
2                 1.0  [{'needsFetchReview': False, 'partnerItemId': ...   
3                 4.0  [{'barcode': '4011', 'description': 'ITEM NOT ...   
4                 2.0  [{'barcode': '4011', 'description': 'ITEM NOT ...   

  rewardsReceiptStatus totalSpent                    userId  
0             FINISHED      26.00  5ff1e1eacfcf6c399c274ae6  
1             FINISHED      11.00  5ff1e194b6a9d73a3a9f1052  
2             REJECTED      10.00  5ff1e1f1cfcf6c399c274b0b  
3             FINISHED      28.00  5ff1e1eacfcf6c399c274ae6  
4             FINISHED       1.00  5ff1e194b6a9d73a3a9f1052

In [4]:
##fixing the id column : extracting the values form dictionaries

receipts["_id"] = receipts["_id"].apply(lambda x : x["$oid"])

In [5]:
receipts.head()

_id  bonusPointsEarned  \
0  5ff1e1eb0a720f0523000575              500.0   
1  5ff1e1bb0a720f052300056b              150.0   
2  5ff1e1f10a720f052300057a                5.0   
3  5ff1e1ee0a7214ada100056f                5.0   
4  5ff1e1d20a7214ada1000561                5.0   

                             bonusPointsEarnedReason  \
0  Receipt number 2 completed, bonus point schedu...   
1  Receipt number 5 completed, bonus point schedu...   
2                         All-receipts receipt bonus   
3                         All-receipts receipt bonus   
4                         All-receipts receipt bonus   

                 createDate               dateScanned  \
0  {'$date': 1609687531000}  {'$date': 1609687531000}   
1  {'$date': 1609687483000}  {'$date': 1609687483000}   
2  {'$date': 1609687537000}  {'$date': 1609687537000}   
3  {'$date': 1609687534000}  {'$date': 1609687534000}   
4  {'$date': 1609687506000}  {'$date': 1609687506000}   

               finishedDate                modifyDate  \
0  {'$date': 1609687531000}  {'$date': 1609687536000}   
1  {'$date': 1609687483000}  {'$date': 1609687488000}   
2                       NaN  {'$date': 1609687542000}   
3  {'$date': 1609687534000}  {'$date': 1609687539000}   
4  {'$date': 1609687511000}  {'$date': 1609687511000}   

          pointsAwardedDate pointsEarned              purchaseDate  \
0  {'$date': 1609687531000}        500.0  {'$date': 1609632000000}   
1  {'$date': 1609687483000}        150.0  {'$date': 1609601083000}   
2                       NaN            5  {'$date': 1609632000000}   
3  {'$date': 1609687534000}          5.0  {'$date': 1609632000000}   
4  {'$date': 1609687506000}          5.0  {'$date': 1609601106000}   

   purchasedItemCount                             rewardsReceiptItemList  \
0                 5.0  [{'barcode': '4011', 'description': 'ITEM NOT ...   
1                 2.0  [{'barcode': '4011', 'description': 'ITEM NOT ...   
2                 1.0  [{'needsFetchReview': False, 'partnerItemId': ...   
3                 4.0  [{'barcode': '4011', 'description': 'ITEM NOT ...   
4                 2.0  [{'barcode': '4011', 'description': 'ITEM NOT ...   

  rewardsReceiptStatus totalSpent                    userId  
0             FINISHED      26.00  5ff1e1eacfcf6c399c274ae6  
1             FINISHED      11.00  5ff1e194b6a9d73a3a9f1052  
2             REJECTED      10.00  5ff1e1f1cfcf6c399c274b0b  
3             FINISHED      28.00  5ff1e1eacfcf6c399c274ae6  
4             FINISHED       1.00  5ff1e194b6a9d73a3a9f1052

In [6]:
'''
this function will fix the date and give unknown if there is no key 
'''

def date_format_fixer(x):
    
    try:
        return( pd.to_datetime(x["$date"], unit='ms'))
    except TypeError:
        return("unknown")

In [7]:
receipts["createDate"] =receipts["createDate"].apply(lambda x : date_format_fixer(x) )
receipts["dateScanned"] =receipts["dateScanned"].apply(lambda x : date_format_fixer(x) )
receipts["finishedDate"] =receipts["finishedDate"].apply(lambda x : date_format_fixer(x) )
receipts["modifyDate"] =receipts["modifyDate"].apply(lambda x : date_format_fixer(x) )
receipts["pointsAwardedDate"] =receipts["pointsAwardedDate"].apply(lambda x : date_format_fixer(x) )
receipts["purchaseDate"] =receipts["purchaseDate"].apply(lambda x : date_format_fixer(x) )

In [8]:
receipts.head()

_id  bonusPointsEarned  \
0  5ff1e1eb0a720f0523000575              500.0   
1  5ff1e1bb0a720f052300056b              150.0   
2  5ff1e1f10a720f052300057a                5.0   
3  5ff1e1ee0a7214ada100056f                5.0   
4  5ff1e1d20a7214ada1000561                5.0   

                             bonusPointsEarnedReason          createDate  \
0  Receipt number 2 completed, bonus point schedu... 2021-01-03 15:25:31   
1  Receipt number 5 completed, bonus point schedu... 2021-01-03 15:24:43   
2                         All-receipts receipt bonus 2021-01-03 15:25:37   
3                         All-receipts receipt bonus 2021-01-03 15:25:34   
4                         All-receipts receipt bonus 2021-01-03 15:25:06   

          dateScanned         finishedDate          modifyDate  \
0 2021-01-03 15:25:31  2021-01-03 15:25:31 2021-01-03 15:25:36   
1 2021-01-03 15:24:43  2021-01-03 15:24:43 2021-01-03 15:24:48   
2 2021-01-03 15:25:37              unknown 2021-01-03 15:25:42   
3 2021-01-03 15:25:34  2021-01-03 15:25:34 2021-01-03 15:25:39   
4 2021-01-03 15:25:06  2021-01-03 15:25:11 2021-01-03 15:25:11   

     pointsAwardedDate pointsEarned         purchaseDate  purchasedItemCount  \
0  2021-01-03 15:25:31        500.0  2021-01-03 00:00:00                 5.0   
1  2021-01-03 15:24:43        150.0  2021-01-02 15:24:43                 2.0   
2              unknown            5  2021-01-03 00:00:00                 1.0   
3  2021-01-03 15:25:34          5.0  2021-01-03 00:00:00                 4.0   
4  2021-01-03 15:25:06          5.0  2021-01-02 15:25:06                 2.0   

                              rewardsReceiptItemList rewardsReceiptStatus  \
0  [{'barcode': '4011', 'description': 'ITEM NOT ...             FINISHED   
1  [{'barcode': '4011', 'description': 'ITEM NOT ...             FINISHED   
2  [{'needsFetchReview': False, 'partnerItemId': ...             REJECTED   
3  [{'barcode': '4011', 'description': 'ITEM NOT ...             FINISHED   
4  [{'barcode': '4011', 'description': 'ITEM NOT ...             FINISHED   

  totalSpent                    userId  
0      26.00  5ff1e1eacfcf6c399c274ae6  
1      11.00  5ff1e194b6a9d73a3a9f1052  
2      10.00  5ff1e1f1cfcf6c399c274b0b  
3      28.00  5ff1e1eacfcf6c399c274ae6  
4       1.00  5ff1e194b6a9d73a3a9f1052

#### Separating the Outer RewardsReceiptItemList into a table and taking receipt number for reference 

In [9]:


all_items = []

# Iterate over 'rewardsReceiptItemList'

for index, itemList in receipts['rewardsReceiptItemList'].dropna().items():
    receipt_id = receipts.loc[index, '_id']  # Get the corresponding receipt ID
    
    # Check if itemList is indeed a list before iterating
    if isinstance(itemList, list):
        for item in itemList:
            if isinstance(item, dict):  # Ensure each item is a dictionary
                item['oid'] = receipt_id  # Assign receipt ID to each item
                all_items.append(item)  # Append the modified item to the list
    else:
        print(f"Warning: Skipping non-list entry at index {index}")

# Create a DataFrame from the list of all items
rewarditemlist = pd.DataFrame(all_items)

In [10]:
rewarditemlist.head()

barcode                                        description finalPrice  \
0          4011                                     ITEM NOT FOUND      26.00   
1          4011                                     ITEM NOT FOUND          1   
2  028400642255  DORITOS TORTILLA CHIP SPICY SWEET CHILI REDUCE...      10.00   
3           NaN                                                NaN        NaN   
4          4011                                     ITEM NOT FOUND      28.00   

  itemPrice needsFetchReview partnerItemId preventTargetGapPoints  \
0     26.00            False             1                   True   
1         1              NaN             1                    NaN   
2     10.00             True             2                   True   
3       NaN            False             1                   True   
4     28.00            False             1                   True   

   quantityPurchased userFlaggedBarcode userFlaggedNewItem  ... itemNumber  \
0                5.0               4011               True  ...        NaN   
1                1.0                NaN                NaN  ...        NaN   
2                1.0       028400642255               True  ...        NaN   
3                NaN               4011               True  ...        NaN   
4                4.0               4011               True  ...        NaN   

   originalMetaBriteQuantityPurchased pointsEarned targetPrice  \
0                                 NaN          NaN         NaN   
1                                 NaN          NaN         NaN   
2                                 NaN          NaN         NaN   
3                                 NaN          NaN         NaN   
4                                 NaN          NaN         NaN   

  competitiveProduct originalFinalPrice originalMetaBriteItemPrice deleted  \
0                NaN                NaN                        NaN     NaN   
1                NaN                NaN                        NaN     NaN   
2                NaN                NaN                        NaN     NaN   
3                NaN                NaN                        NaN     NaN   
4                NaN                NaN                        NaN     NaN   

  priceAfterCoupon metabriteCampaignId  
0              NaN                 NaN  
1              NaN                 NaN  
2              NaN                 NaN  
3              NaN                 NaN  
4              NaN                 NaN  

[5 rows x 35 columns]

In [11]:
receipts = receipts.drop(columns = ["rewardsReceiptItemList"],axis = 1)

In [12]:
receipts.head()

_id  bonusPointsEarned  \
0  5ff1e1eb0a720f0523000575              500.0   
1  5ff1e1bb0a720f052300056b              150.0   
2  5ff1e1f10a720f052300057a                5.0   
3  5ff1e1ee0a7214ada100056f                5.0   
4  5ff1e1d20a7214ada1000561                5.0   

                             bonusPointsEarnedReason          createDate  \
0  Receipt number 2 completed, bonus point schedu... 2021-01-03 15:25:31   
1  Receipt number 5 completed, bonus point schedu... 2021-01-03 15:24:43   
2                         All-receipts receipt bonus 2021-01-03 15:25:37   
3                         All-receipts receipt bonus 2021-01-03 15:25:34   
4                         All-receipts receipt bonus 2021-01-03 15:25:06   

          dateScanned         finishedDate          modifyDate  \
0 2021-01-03 15:25:31  2021-01-03 15:25:31 2021-01-03 15:25:36   
1 2021-01-03 15:24:43  2021-01-03 15:24:43 2021-01-03 15:24:48   
2 2021-01-03 15:25:37              unknown 2021-01-03 15:25:42   
3 2021-01-03 15:25:34  2021-01-03 15:25:34 2021-01-03 15:25:39   
4 2021-01-03 15:25:06  2021-01-03 15:25:11 2021-01-03 15:25:11   

     pointsAwardedDate pointsEarned         purchaseDate  purchasedItemCount  \
0  2021-01-03 15:25:31        500.0  2021-01-03 00:00:00                 5.0   
1  2021-01-03 15:24:43        150.0  2021-01-02 15:24:43                 2.0   
2              unknown            5  2021-01-03 00:00:00                 1.0   
3  2021-01-03 15:25:34          5.0  2021-01-03 00:00:00                 4.0   
4  2021-01-03 15:25:06          5.0  2021-01-02 15:25:06                 2.0   

  rewardsReceiptStatus totalSpent                    userId  
0             FINISHED      26.00  5ff1e1eacfcf6c399c274ae6  
1             FINISHED      11.00  5ff1e194b6a9d73a3a9f1052  
2             REJECTED      10.00  5ff1e1f1cfcf6c399c274b0b  
3             FINISHED      28.00  5ff1e1eacfcf6c399c274ae6  
4             FINISHED       1.00  5ff1e194b6a9d73a3a9f1052

#### Now Brands 

In [13]:

brands.head()

_id       barcode        category  \
0  {'$oid': '601ac115be37ce2ead437551'}  511111019862          Baking   
1  {'$oid': '601c5460be37ce2ead43755f'}  511111519928       Beverages   
2  {'$oid': '601ac142be37ce2ead43755d'}  511111819905          Baking   
3  {'$oid': '601ac142be37ce2ead43755a'}  511111519874          Baking   
4  {'$oid': '601ac142be37ce2ead43755e'}  511111319917  Candy & Sweets   

       categoryCode                                                cpg  \
0            BAKING  {'$id': {'$oid': '601ac114be37ce2ead437550'}, ...   
1         BEVERAGES  {'$id': {'$oid': '5332f5fbe4b03c9a25efd0ba'}, ...   
2            BAKING  {'$id': {'$oid': '601ac142be37ce2ead437559'}, ...   
3            BAKING  {'$id': {'$oid': '601ac142be37ce2ead437559'}, ...   
4  CANDY_AND_SWEETS  {'$id': {'$oid': '5332fa12e4b03c9a25efd1e7'}, ...   

                        name topBrand                      brandCode  
0  test brand @1612366101024    False                            NaN  
1                  Starbucks    False                      STARBUCKS  
2  test brand @1612366146176    False  TEST BRANDCODE @1612366146176  
3  test brand @1612366146051    False  TEST BRANDCODE @1612366146051  
4  test brand @1612366146827    False  TEST BRANDCODE @1612366146827

In [14]:
## we can use json normalize as there are more than one field in the dictionary

brands_norm = pd.json_normalize(brands['cpg'])

In [15]:
brands_norm.head(10)

$ref                  $id.$oid
0  Cogs  601ac114be37ce2ead437550
1  Cogs  5332f5fbe4b03c9a25efd0ba
2  Cogs  601ac142be37ce2ead437559
3  Cogs  601ac142be37ce2ead437559
4  Cogs  5332fa12e4b03c9a25efd1e7
5  Cogs  601ac142be37ce2ead437559
6  Cogs  601ac142be37ce2ead437559
7  Cogs  559c2234e4b06aca36af13c6
8  Cogs  5a734034e4b0d58f376be874
9  Cogs  59ba6f1ce4b092b29c167346

In [16]:
brands_norm = brands_norm.add_prefix('cpg.') 

In [17]:
brands_norm

cpg.$ref              cpg.$id.$oid
0        Cogs  601ac114be37ce2ead437550
1        Cogs  5332f5fbe4b03c9a25efd0ba
2        Cogs  601ac142be37ce2ead437559
3        Cogs  601ac142be37ce2ead437559
4        Cogs  5332fa12e4b03c9a25efd1e7
...       ...                       ...
1162     Cogs  5f77274dbe37ce6b592e90bf
1163     Cogs  53e10d6368abd3c7065097cc
1164     Cogs  5332fa12e4b03c9a25efd1e7
1165     Cogs  5332f5f6e4b03c9a25efd0b4
1166     Cogs  6026d757be37ce6369301467

[1167 rows x 2 columns]

In [18]:
brands_final = pd.merge(brands, brands_norm, left_index=True, right_index=True, how='outer')   

In [19]:
brands_final.head()

_id       barcode        category  \
0  {'$oid': '601ac115be37ce2ead437551'}  511111019862          Baking   
1  {'$oid': '601c5460be37ce2ead43755f'}  511111519928       Beverages   
2  {'$oid': '601ac142be37ce2ead43755d'}  511111819905          Baking   
3  {'$oid': '601ac142be37ce2ead43755a'}  511111519874          Baking   
4  {'$oid': '601ac142be37ce2ead43755e'}  511111319917  Candy & Sweets   

       categoryCode                                                cpg  \
0            BAKING  {'$id': {'$oid': '601ac114be37ce2ead437550'}, ...   
1         BEVERAGES  {'$id': {'$oid': '5332f5fbe4b03c9a25efd0ba'}, ...   
2            BAKING  {'$id': {'$oid': '601ac142be37ce2ead437559'}, ...   
3            BAKING  {'$id': {'$oid': '601ac142be37ce2ead437559'}, ...   
4  CANDY_AND_SWEETS  {'$id': {'$oid': '5332fa12e4b03c9a25efd1e7'}, ...   

                        name topBrand                      brandCode cpg.$ref  \
0  test brand @1612366101024    False                            NaN     Cogs   
1                  Starbucks    False                      STARBUCKS     Cogs   
2  test brand @1612366146176    False  TEST BRANDCODE @1612366146176     Cogs   
3  test brand @1612366146051    False  TEST BRANDCODE @1612366146051     Cogs   
4  test brand @1612366146827    False  TEST BRANDCODE @1612366146827     Cogs   

               cpg.$id.$oid  
0  601ac114be37ce2ead437550  
1  5332f5fbe4b03c9a25efd0ba  
2  601ac142be37ce2ead437559  
3  601ac142be37ce2ead437559  
4  5332fa12e4b03c9a25efd1e7

In [20]:
## to fix the id column

brands_final['_id'] = brands_final['_id'].apply(lambda x: x['$oid'])

In [21]:
brands_final.head()

_id       barcode        category      categoryCode  \
0  601ac115be37ce2ead437551  511111019862          Baking            BAKING   
1  601c5460be37ce2ead43755f  511111519928       Beverages         BEVERAGES   
2  601ac142be37ce2ead43755d  511111819905          Baking            BAKING   
3  601ac142be37ce2ead43755a  511111519874          Baking            BAKING   
4  601ac142be37ce2ead43755e  511111319917  Candy & Sweets  CANDY_AND_SWEETS   

                                                 cpg  \
0  {'$id': {'$oid': '601ac114be37ce2ead437550'}, ...   
1  {'$id': {'$oid': '5332f5fbe4b03c9a25efd0ba'}, ...   
2  {'$id': {'$oid': '601ac142be37ce2ead437559'}, ...   
3  {'$id': {'$oid': '601ac142be37ce2ead437559'}, ...   
4  {'$id': {'$oid': '5332fa12e4b03c9a25efd1e7'}, ...   

                        name topBrand                      brandCode cpg.$ref  \
0  test brand @1612366101024    False                            NaN     Cogs   
1                  Starbucks    False                      STARBUCKS     Cogs   
2  test brand @1612366146176    False  TEST BRANDCODE @1612366146176     Cogs   
3  test brand @1612366146051    False  TEST BRANDCODE @1612366146051     Cogs   
4  test brand @1612366146827    False  TEST BRANDCODE @1612366146827     Cogs   

               cpg.$id.$oid  
0  601ac114be37ce2ead437550  
1  5332f5fbe4b03c9a25efd0ba  
2  601ac142be37ce2ead437559  
3  601ac142be37ce2ead437559  
4  5332fa12e4b03c9a25efd1e7

### finallly the user table 


In [22]:
users.head()

_id  active               createdDate  \
0  {'$oid': '5ff1e194b6a9d73a3a9f1052'}    True  {'$date': 1609687444800}   
1  {'$oid': '5ff1e194b6a9d73a3a9f1052'}    True  {'$date': 1609687444800}   
2  {'$oid': '5ff1e194b6a9d73a3a9f1052'}    True  {'$date': 1609687444800}   
3  {'$oid': '5ff1e1eacfcf6c399c274ae6'}    True  {'$date': 1609687530554}   
4  {'$oid': '5ff1e194b6a9d73a3a9f1052'}    True  {'$date': 1609687444800}   

                  lastLogin      role signUpSource state  
0  {'$date': 1609687537858}  consumer        Email    WI  
1  {'$date': 1609687537858}  consumer        Email    WI  
2  {'$date': 1609687537858}  consumer        Email    WI  
3  {'$date': 1609687530597}  consumer        Email    WI  
4  {'$date': 1609687537858}  consumer        Email    WI

In [23]:
## for the id column

users['_id'] = users['_id'].apply(lambda x: x['$oid'])

In [24]:
## fixing the date columns using the same function

users['createdDate'] = users['createdDate'].apply(lambda x: date_format_fixer(x))
users['lastLogin'] = users['lastLogin'].apply(lambda x: date_format_fixer(x))

In [25]:
users.head()

_id  active             createdDate  \
0  5ff1e194b6a9d73a3a9f1052    True 2021-01-03 15:24:04.800   
1  5ff1e194b6a9d73a3a9f1052    True 2021-01-03 15:24:04.800   
2  5ff1e194b6a9d73a3a9f1052    True 2021-01-03 15:24:04.800   
3  5ff1e1eacfcf6c399c274ae6    True 2021-01-03 15:25:30.554   
4  5ff1e194b6a9d73a3a9f1052    True 2021-01-03 15:24:04.800   

                    lastLogin      role signUpSource state  
0  2021-01-03 15:25:37.858000  consumer        Email    WI  
1  2021-01-03 15:25:37.858000  consumer        Email    WI  
2  2021-01-03 15:25:37.858000  consumer        Email    WI  
3  2021-01-03 15:25:30.597000  consumer        Email    WI  
4  2021-01-03 15:25:37.858000  consumer        Email    WI

# exporting and csv's

In [26]:

### for ease of export to mysql workbench ( it drops the record where numerical column values are null so taking 0 )

receipts['bonusPointsEarned'] = receipts['bonusPointsEarned'].fillna(0.00)
receipts['pointsEarned'] = receipts['pointsEarned'].fillna(0.00)
receipts['purchasedItemCount'] = receipts['purchasedItemCount'].fillna(0.00)
receipts['totalSpent'] = receipts['totalSpent'].fillna(0.00)


rewarditemlist["quantityPurchased"] = rewarditemlist["quantityPurchased"].fillna(0.00)


In [27]:
## coverting dataframes to csv for further sql querying 

# brands_final.to_csv("brands_final.csv")
# receipts.to_csv("receipts_final.csv")
# users.to_csv("users.csv")
# rewarditemlist.to_csv("rewarditemlist.csv")


## Data Quality Issues

Reason for a weak join between Receipts and Brands :
Mismatch of Brandcodes in both tables i.e. Some brandcodes present in brands but missing in Receipts is not an issue, but brands present in receipts and missing in brands table is a data quality issue.

In [28]:
brands_final["brandCode"].nunique()                    ## number of unique brands in brands tables including nulls

897

In [29]:
rewarditemlist["brandCode"].nunique()     ## number of unique brands in receipts table

227

In [30]:
receipts_brandcode = rewarditemlist["brandCode"].unique()
brands_brandcode = brands_final['brandCode'].unique()

In [31]:
exists_in_brands = pd.Series(receipts_brandcode).isin(brands_brandcode)

In [32]:
rewarditemlist["brandCode"].nunique() - exists_in_brands.sum() 

185

#### 185 brands( out of 227) in Receipt table are missing in Brands table ( therefore barcode and brandcode joins will have a lot of data quality issues) 

### Duplications in Users 

In [33]:
users.head()

_id  active             createdDate  \
0  5ff1e194b6a9d73a3a9f1052    True 2021-01-03 15:24:04.800   
1  5ff1e194b6a9d73a3a9f1052    True 2021-01-03 15:24:04.800   
2  5ff1e194b6a9d73a3a9f1052    True 2021-01-03 15:24:04.800   
3  5ff1e1eacfcf6c399c274ae6    True 2021-01-03 15:25:30.554   
4  5ff1e194b6a9d73a3a9f1052    True 2021-01-03 15:24:04.800   

                    lastLogin      role signUpSource state  
0  2021-01-03 15:25:37.858000  consumer        Email    WI  
1  2021-01-03 15:25:37.858000  consumer        Email    WI  
2  2021-01-03 15:25:37.858000  consumer        Email    WI  
3  2021-01-03 15:25:30.597000  consumer        Email    WI  
4  2021-01-03 15:25:37.858000  consumer        Email    WI

In [34]:
users.shape

(495, 7)

In [35]:
users["_id"].nunique()

212

#### WE have only 212 unique user ids

In [36]:
users[users["_id"] == "54943462e4b07e684157a532"].reset_index()  ### example to show duplicates

index                       _id  active             createdDate  \
0     475  54943462e4b07e684157a532    True 2014-12-19 14:21:22.381   
1     476  54943462e4b07e684157a532    True 2014-12-19 14:21:22.381   
2     477  54943462e4b07e684157a532    True 2014-12-19 14:21:22.381   
3     478  54943462e4b07e684157a532    True 2014-12-19 14:21:22.381   
4     479  54943462e4b07e684157a532    True 2014-12-19 14:21:22.381   
5     480  54943462e4b07e684157a532    True 2014-12-19 14:21:22.381   
6     481  54943462e4b07e684157a532    True 2014-12-19 14:21:22.381   
7     482  54943462e4b07e684157a532    True 2014-12-19 14:21:22.381   
8     483  54943462e4b07e684157a532    True 2014-12-19 14:21:22.381   
9     484  54943462e4b07e684157a532    True 2014-12-19 14:21:22.381   
10    485  54943462e4b07e684157a532    True 2014-12-19 14:21:22.381   
11    486  54943462e4b07e684157a532    True 2014-12-19 14:21:22.381   
12    487  54943462e4b07e684157a532    True 2014-12-19 14:21:22.381   
13    488  54943462e4b07e684157a532    True 2014-12-19 14:21:22.381   
14    489  54943462e4b07e684157a532    True 2014-12-19 14:21:22.381   
15    490  54943462e4b07e684157a532    True 2014-12-19 14:21:22.381   
16    491  54943462e4b07e684157a532    True 2014-12-19 14:21:22.381   
17    492  54943462e4b07e684157a532    True 2014-12-19 14:21:22.381   
18    493  54943462e4b07e684157a532    True 2014-12-19 14:21:22.381   
19    494  54943462e4b07e684157a532    True 2014-12-19 14:21:22.381   

                     lastLogin         role signUpSource state  
0   2021-03-05 16:52:23.204000  fetch-staff          NaN   NaN  
1   2021-03-05 16:52:23.204000  fetch-staff          NaN   NaN  
2   2021-03-05 16:52:23.204000  fetch-staff          NaN   NaN  
3   2021-03-05 16:52:23.204000  fetch-staff          NaN   NaN  
4   2021-03-05 16:52:23.204000  fetch-staff          NaN   NaN  
5   2021-03-05 16:52:23.204000  fetch-staff          NaN   NaN  
6   2021-03-05 16:52:23.204000  fetch-staff          NaN   NaN  
7   2021-03-05 16:52:23.204000  fetch-staff          NaN   NaN  
8   2021-03-05 16:52:23.204000  fetch-staff          NaN   NaN  
9   2021-03-05 16:52:23.204000  fetch-staff          NaN   NaN  
10  2021-03-05 16:52:23.204000  fetch-staff          NaN   NaN  
11  2021-03-05 16:52:23.204000  fetch-staff          NaN   NaN  
12  2021-03-05 16:52:23.204000  fetch-staff          NaN   NaN  
13  2021-03-05 16:52:23.204000  fetch-staff          NaN   NaN  
14  2021-03-05 16:52:23.204000  fetch-staff          NaN   NaN  
15  2021-03-05 16:52:23.204000  fetch-staff          NaN   NaN  
16  2021-03-05 16:52:23.204000  fetch-staff          NaN   NaN  
17  2021-03-05 16:52:23.204000  fetch-staff          NaN   NaN  
18  2021-03-05 16:52:23.204000  fetch-staff          NaN   NaN  
19  2021-03-05 16:52:23.204000  fetch-staff          NaN   NaN

In [37]:
brands_final.head()

_id       barcode        category      categoryCode  \
0  601ac115be37ce2ead437551  511111019862          Baking            BAKING   
1  601c5460be37ce2ead43755f  511111519928       Beverages         BEVERAGES   
2  601ac142be37ce2ead43755d  511111819905          Baking            BAKING   
3  601ac142be37ce2ead43755a  511111519874          Baking            BAKING   
4  601ac142be37ce2ead43755e  511111319917  Candy & Sweets  CANDY_AND_SWEETS   

                                                 cpg  \
0  {'$id': {'$oid': '601ac114be37ce2ead437550'}, ...   
1  {'$id': {'$oid': '5332f5fbe4b03c9a25efd0ba'}, ...   
2  {'$id': {'$oid': '601ac142be37ce2ead437559'}, ...   
3  {'$id': {'$oid': '601ac142be37ce2ead437559'}, ...   
4  {'$id': {'$oid': '5332fa12e4b03c9a25efd1e7'}, ...   

                        name topBrand                      brandCode cpg.$ref  \
0  test brand @1612366101024    False                            NaN     Cogs   
1                  Starbucks    False                      STARBUCKS     Cogs   
2  test brand @1612366146176    False  TEST BRANDCODE @1612366146176     Cogs   
3  test brand @1612366146051    False  TEST BRANDCODE @1612366146051     Cogs   
4  test brand @1612366146827    False  TEST BRANDCODE @1612366146827     Cogs   

               cpg.$id.$oid  
0  601ac114be37ce2ead437550  
1  5332f5fbe4b03c9a25efd0ba  
2  601ac142be37ce2ead437559  
3  601ac142be37ce2ead437559  
4  5332fa12e4b03c9a25efd1e7

In [38]:
brands_final.shape

(1167, 10)

In [39]:
brands_final["_id"].nunique() ## looks fine

1167